src/context.cc

```diff
@@ -127,17 +127,28 @@ class SharedData {
     it->second.queue.pop_front();
     return WasmResult::Ok;
   }
+
   WasmResult enqueue(uint32_t token, string_view value) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    auto it = queues_.find(token);
-    if (it == queues_.end()) {
-      return WasmResult::NotFound;
+    Queue *target_queue;
+    std::string vm_key;
+    uint32_t context_id;
+
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      auto it = queues_.find(token);
+      if (it == queues_.end()) {
+        return WasmResult::NotFound;
+      }
+      target_queue = &(it->second);
+      vm_key = target_queue->vm_key;
+      context_id = target_queue->context_id;
+      target_queue->queue.push_back(std::string(value));
     }
-    it->second.queue.push_back(std::string(value));
-    auto vm_id = it->second.vm_id;
-    auto context_id = it->second.context_id;
-    it->second.call_on_thread([vm_id, context_id, token] {
-      auto wasm = getThreadLocalWasm(vm_id);
+
+    target_queue->call_on_thread([vm_key, context_id, token] {
```

Comment:
You are holding the queue outside the lock now which isn't safe as that object can be mutated in registerQueue on a different thread.  Either copy the call_on_thread function under the lock and use that or make the call under the lock as it was before.

Reply:
Agreed! I pushed a fix.